### DISCUSSION TITLE
Using python's super() within a new Julia-defined Python class

### DISCUSSION THREAD

**Discussion by bensetterholm** [cjdoris/PythonCall.jl#423] [Q&A] [answered]
I need to define a Python class that inherits another Python class with Python's `super()` function in my Julia code. However, I don't know how to access `super` (presuming that is currently even possible).

As a MWE of my problem, here is an attempt to translate the following pure Python code (taken from https://realpython.com/python-super/) into its equivalent PythonCall representation.

Python:
```python
class Rectangle:
    def __init__(self, length, width):
        self.length = length
        self.width = width

    def area(self):
        return self.length * self.width

    def perimeter(self):
        return 2 * self.length + 2 * self.width

class Square(Rectangle):
    def __init__(self, length):
        super().__init__(length, length)
```

Julia attempt:
```julia
using PythonCall

Rectangle = pytype("Rectangle", (), [
    "__module__" => "__main__",

    pyfunc(
        name = "__init__",
        function (self, length, width)
            self.length = length
            self.width = width
            return
        end
    ),

    pyfunc(
        name = "area",
        self -> self.length * self.width
    ),

    pyfunc(
        name = "perimeter",
        self -> 2 * self.length + 2 * self.width
    ),
])

Square = pytype("Square", (Rectangle,), [
    "__module__" => "__main__",

    pyfunc(
        name = "__init__",
        function (self, length)
            super().__init__(length, length)
            return
        end
    )
])
```

Note that `Rectangle` works exactly as expected. However `Square` does not work due to the undefined `super()`. I do not see anything like `pysuper()` defined in PythonCall either. Am I missing something, or is this currently not possible?

**Comment by cjdoris** [comment, marked as answer]
Not currently possible I'm afraid. You could do `Rectangle.__init__(self, ...)` instead.

**Reply by bensetterholm** [reply]
Thank you; this seems to do the trick! Are there any potential "gotchas" you are aware of to using this syntax instead of `super`? (I only write Python code when I absolutely have to, and don't necessarily have a good mental model of all of the nuances of Python classes.)

**Reply by cjdoris** [reply]
If your class only has one base class then I think both ways are equivalent, whereas `super()` can properly handle multiple bases by traversing the MRO properly.
